src/misc/lv_array.c

```diff
@@ -45,86 +45,128 @@ void lv_array_init(lv_array_t * array, uint32_t capacity, uint32_t element_size)
 {
-    if(lv_array_is_empty(array)) {
+    if(lv_array_is_empty(source)) {
         return;
     }
-    lv_array_destroy(target);
-    lv_array_init(target, array->capacity, array->element_size);
-    lv_memcpy(target->data, array->data, array->size * array->element_size);
-    target->size = array->size;
+    lv_array_deinit(target);
+    lv_array_init(target, source->capacity, source->element_size);
+    lv_memcpy(target->data, source->data, source->size * source->element_size);
+    target->size = source->size;
 }
 
-void lv_array_clear(lv_array_t * array)
+lv_result_t lv_array_remove(lv_array_t * array, uint32_t index)
 {
-    array->size = 0;
+    if(index >= array->size) {
+        return LV_RESULT_INVALID;
+    }
+
+    /*Shortcut*/
+    if(index == array->size - 1) {
+        lv_array_resize(array, array->size - 1);
+        return LV_RESULT_OK;
+    }
+
+    uint8_t * start = lv_array_at(array, index);
+    uint8_t * remaining = start + array->element_size;
+    uint32_t remaining_size = (array->size - index - 1) * array->element_size;
+    lv_memcpy(start, remaining, remaining_size);
+    lv_array_resize(array, array->size - 1);
+    return LV_RESULT_OK;
 }
 
-void lv_array_resize(lv_array_t * array, uint32_t new_capacity)
+lv_result_t lv_array_erase(lv_array_t * array, uint32_t start, uint32_t end)
 {
-    if(new_capacity > array->size) {
-        if(new_capacity > array->capacity) {
-            uint8_t * data = lv_malloc(new_capacity * array->element_size);
-            lv_memcpy(data, array->data, array->size * array->element_size);
-            lv_free(array->data);
-            array->data = data;
-            array->capacity = new_capacity;
-        }
+    if(end > array->size) {
+        end = array->size;
     }
-    else {
+
+    if(start >= end) {
+        return LV_RESULT_INVALID;
+    }
+
+    /*Shortcut*/
+    if(end == array->size) {
+        lv_array_resize(array, start);
+        return LV_RESULT_OK;
+    }
+
+    uint8_t * start_p = lv_array_at(array, start);
+    uint8_t * remaining = start_p + (end - start) * array->element_size;
+    uint32_t remaining_size = (array->size - end) * array->element_size;
+    lv_memcpy(start_p, remaining, remaining_size);
+    lv_array_resize(array, array->size - (end - start));
+    return LV_RESULT_OK;
+}
+
+void lv_array_resize(lv_array_t * array, uint32_t new_capacity)
+{
+    uint8_t * data = lv_realloc(array->data, new_capacity * array->element_size);
+    LV_ASSERT_NULL(data);
+    array->data = data;
+    array->capacity = new_capacity;
+    if(array->size > new_capacity) {
         array->size = new_capacity;
     }
 }
 
-bool lv_array_append(lv_array_t * array, const uint8_t * element)
+lv_result_t lv_array_concat(lv_array_t * array, const lv_array_t * other)
 {
-    if(array->size >= array->capacity) {
-        // array is full
-        return false;
+    LV_ASSERT_NULL(array->data);
+    uint32_t size = other->size;
+    if(array->size + size > array->capacity) {
```

Comment:
Maybe we should allow the array to auto resize.

I think whenever we use this lv_array, we don't want to care about whether it is full or not, we just add elements, concat arrays, get element at given index, remove element, etc.... like in any other programming language (for example JavaScript).... so what I want to say we just want to use it like a "service", and don't want to pre-check the capacity every time we want to add a new element or concat another array. It's not the caller's responsibility.
I don't see a reason why we should reject array concat when the target is full.

Or is this behaviour of this concat function intentionally?

PS:
I see, that `lv_array_push_back()` automatically resizes the array.

Reply:
The array module doesn't have auto-resize capability originally. But since push_back does in this way, concat should too. Updated.

As for the performance issue when removing element, it's a tradeoff between size and efficiency.
I cannot decide what to do yet in this PR. It would be better for a new issue to discuss.

Reply:
I think it is OK now, even removing element.
If there will be a performance issue, because of many push and removes (continous resize and memory allocation), then we could think on having a minimum capacity threshold, so if remove function is called, then it will not resize the array if size is smaller than the threshold.

Reply:
I want to keep the  old API behavior same as the original one.
The newly introduced API does have auto resize now.

Reply:
Great, thanks, it now works perfectly (as arrays in other prog.languages)
When you think it is done, and CI builds finished successfully, could you please re-request review? So I can mark our conversation comments resolved. Thank you!
![image](https://github.com/lvgl/lvgl/assets/25751703/786928e3-b7de-4933-92c5-91332ade13c2)

PS: 
it is recommended to re-request review from **all** previous reviewers to review again the whole PR (or just recent commits).